File: Data Structures & Algorithms/daily-temperatures/submission-0.java
class Solution {
    public int[] dailyTemperatures(int[] temp) {
        int n = temp.length;

        int[] ans = new int[n];
        ArrayDeque<Integer> stack = new ArrayDeque();

        for (int i = 0; i < n; i++) {
            int curr = temp[i];

            while (!stack.isEmpty() && curr > temp[stack.peek()]) {
                int idx = stack.pop();
                ans[idx] = i - idx; // distance
            }

            stack.push(i);
        }

        return ans;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        stack = []
        ans = 0

        for i in range(n):
            curr = heights[i]

            while stack and curr < heights[stack[-1]]:
                height = heights[stack.pop()]
                left = -1
                if stack:
                    left = stack[-1]
                
                width = i - 1 - left
                
                ans = max(ans, height * width)
        
            stack.append(i)

        while stack:
            height = heights[stack.pop()]
            left = -1
            if stack:
                left = stack[-1]
            
            width = n - 1 - left
            
            ans = max(ans, height * width)

        return ans



